### PR TITLE
[NUI] Add Wearable.Popup

### PR DESCRIPTION
**Pull request by dongsug-song**
### Description of Change ###
[WIP][NUI] Add Wearable.Popup

- initial commit of Wearable.Popup
- work in progress

### API Changes ###
will be done after review completed

### CODE REVIEW

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+{
+    using L = Tizen.Log;
+
+    /// <summary>
+    /// Popup is an UI component to give a notification or message, which interaction with user simply.
+    /// It is attached to Window directly so that it is shown on top of all UI components.
+    /// Title(text) and Content(container) are initially formed and user can control them by setting properties or styles.
+    /// </summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public class Popup : Control
```
Comment (hyunjushin) [inline src/Tizen.NUI.Wearable/src/public/Popup.cs:33]: Are there many differences from NUI.Component.Popup?

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+                contentContainer.WidthSpecification = LayoutParamPolicies.WrapContent;
+                contentContainer.HeightSpecification = LayoutParamPolicies.WrapContent;
+            }
+        }
+
+        /// <summary>
+        /// PopupAnimation
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public interface PopupAnimation
```
Comment (hyunjushin) [inline src/Tizen.NUI.Wearable/src/public/Popup.cs:300]: Button used Extension for animation customization. Which design is better?

File: src/Tizen.NUI.Wearable/src/public/WearableStyle/PopupStyle.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        static PopupStyle() { }
+
+        /// <summary>
+        /// Creates a new instance of a PopupStyle.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public PopupStyle() : base()
+        {
+            initSubStyle();
```
Comment (hyunjushin) [inline src/Tizen.NUI.Wearable/src/public/WearableStyle/PopupStyle.cs:38]: What's the meaning of SubStyle?

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+        {
+            set => wrapContent = value;
+            get => wrapContent;
+        }
+
+        /// <summary>
+        /// Append Button created by user.
+        /// </summary>
+        /// <remarks>
+        /// Do not use LeftButton, RightButton, SingleButton as index string when using prebuilt types.
```
[the remark this answers is not in the document]
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs]: it is wrong comments. it has been fixed and updated.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs (hunk 1 above)
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs:33]: I tried to change this popup to be inherited from Components.Popup. However, it has several things to be discussed. 

1. Components.Popup uses ButtonGroup and it's visual update is done in ButtonGroup class, and Popup also has private method to update it's own visuality. To make this enable, they should be changed as virtual methods and it looks not very intuitive to override two depth of instance. 

2. AddButton() interface doesn't have the positioning or anchoring of button, so it can't be overrided this method. also this method will be appeared by inheriting, so that users will be confused. 

3. It is possible to implement by inheriting Popup, but the implementation will be overlapped and it seems to have no merit to inherit.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs (hunk 2 above)
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs:344]: I think Button's Extension has more functionality and tuning points. but I think popup needs only "show" and "hide" animations. it seems that popup has only simple behaviors of showing and hiding.

File: src/Tizen.NUI.Wearable/src/public/WearableStyle/PopupStyle.cs (hunk 3 above)
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/WearableStyle/PopupStyle.cs:38]: I followed other Style class naming examples. I think it's because popup style inherits from control style. so it means child class's initializing, I think...

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+
+            contentContainer = new View()
+            {
+                Layout = scrollLayout,
+                PositionUsesPivotPoint = true,
+                ParentOrigin = Tizen.NUI.ParentOrigin.TopCenter,
+                PivotPoint = Tizen.NUI.PivotPoint.TopCenter,
+            };
+
+            scroll = new ScrollableBase()
```
Comment (neostom432) [inline src/Tizen.NUI.Wearable/src/public/Popup.cs:474]: If you want to divide popup into two area like below, you can use LinearLayout .
ㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡ
|⠀⠀⠀⠀⠀⠀title⠀⠀⠀⠀⠀⠀⠀⠀⠀|
ㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡ
|⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀|
|⠀⠀⠀ScrollableBase⠀⠀⠀⠀⠀|
|⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀|
ㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡㅡ

If you set size of title and give weight 1 to ScrollableBase, the size of ScrollableBase is always popup size - title size.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+        {
+            if (previousState == PointStateType.Down && e.Touch.GetState(0) == PointStateType.Finished)
+            {
+                OutsideClicked?.Invoke(this, null);
+            }
+            previousState = e.Touch.GetState(0);
+            return true;
+        }
+
+        private void defaultShowAnimation()
```
Comment (neostom432) [inline src/Tizen.NUI.Wearable/src/public/Popup.cs:533]: I think target of show/hide animation is not popup.
It should be container of popup because popup can be smaller than screen.

but in tizen wearable case, popup is always cover whole screen so... maybe it works.

File: test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/DaliDemo/DaliTableView.cs [hunk 7]
```diff
@@ -133,14 +133,16 @@ public void Initialize()
             logo.ParentOrigin = new Position(0.5f, 0.1f, 0.5f);
             logo.WidthResizePolicy = ResizePolicyType.UseNaturalSize;
             logo.HeightResizePolicy = ResizePolicyType.UseNaturalSize;
+            float logoScale = (float)(NUIApplication.GetDefaultWindow().Size.Height) / 1080.0f;
```
Comment (neostom432) [inline test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/DaliDemo/DaliTableView.cs:136]: This change is not related to Popup.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+        /// <param name="position">Position on popup.</param>
+        public void AppendButton(string index, Button button, ButtonPosition position = ButtonPosition.Custom)
+        {
+            if (button != null)
+            {
+                L.Debug(tag, "AppendButton()");
+                buttonList.Add(index, button);
+                Add(button);
+                button.RaiseToTop();
+                button.Style.Overlay.SiblingOrder = 100; //will be fixed.
```
[the remark this answers is not in the document]
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs]: I tried to change it as using Layout, but it is still not shown. I will check continuously.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+                contentContainer.Size = new Size(width, height);
+                wrapContent = false;
+            }
+        }
+
+        /// <summary>
+        /// OnUpdate
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        protected override void OnUpdate()
```
[the remark this answers is not in the document]
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs:388]: I have changed ContentContainer as public

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)] 
+        public PopupAnimation CustomAnimation;
+
+
+        /// <summary>
+        /// Event when a Button clicked in Popup.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public event EventHandler ButtonClicked
```
[the remark this answers is not in the document]
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs]: I removed it

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+        /// Event when outside of components(Button, Content, Title) is clicked.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public event EventHandler OutsideClicked;
+
+        /// <summary>
+        /// Event when Post method is called. Customized action such as animation can be added.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public event EventHandler Posting;
```
[the remark this answers is not in the document]
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs]: I added BeforePosting, AfterPosting, BeforeDissmising, AfterDissmising events.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+
+        private void initialize()
+        {
+            popupStyle = Style as PopupStyle;
+            if (popupStyle == null)
+            {
+                throw new Exception("Popup style is null.");
+            }
+
+            instanceCnt++;
```
[the remark this answers is not in the document]
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs]: I have removed it.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+                ScrollEnabled = true,
+                WidthResizePolicy = ResizePolicyType.FillToParent,
+                HeightResizePolicy = ResizePolicyType.FillToParent,
+            };
+            Add(scroll);
+            scroll.Add(contentContainer);
+
+            scrollList = new Dictionary<string, View>();
+
+            TouchEvent += Popup_TouchEvent;
```
[the remark this answers is not in the document]
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs:531]: layer doesn't have TouchEvent, so I added it in popup. I think it would be OK that popup can cover all layer and handle all TouchEvents.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs (hunk 5 above)
Comment (dongsug-song) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs:512]: I tried to change it as using Layout, but it is still not shown. I will check continuously.

File: test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/DaliDemo/DaliTableView.cs (hunk 7 above)
Comment (dongsug-song) [reply test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/DaliDemo/DaliTableView.cs:136]: yes, absolutely right. but if this is not changed, all texts in sample app are not shown. (they shows as points => ... )
please accept this change. thank you.

**Comment by dongsug-song** [comment]
@neostom432 
Hello, please review this again. 
The modification of changing to use Layout is not working, so I left it as previous codes.
I will check it continually, so except for this, please review again.
thank you.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs (hunk 5 above)
Comment (neostom432) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs:512]: Please don't forget using Layout later.

File: src/Tizen.NUI.Wearable/src/public/Popup.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (neostom432) [reply src/Tizen.NUI.Wearable/src/public/Popup.cs:531]: It is okay when popup is covered all screen like wearable, but most other case there is a 
transparent background which contains popup. Many apps allows to user dismiss popup touching that area.

So, It is WearablePopup, it is fine but if you have any plan to move this to Tizen.NUI.Components , I think we should make modal components and remove this feature from popup.
